File: sem5/CN/crc.c
#include <stdio.h>
#include <string.h>

int main() {
    int choice;
    char data_str[100], gen_str[50];
    int data[150], gen[50];
    int data_len, gen_len, total_len;
    int i, j;

    printf("1. Generate CRC Code\n");
    printf("2. Verify CRC Code\n");
    printf("Enter your choice: ");
    scanf("%d", &choice);

    switch (choice) {
        case 1:
            printf("\n-- CRC Generation --\n");
            printf("Enter data bits: ");
            scanf("%s", data_str);
            printf("Enter generator polynomial: ");
            scanf("%s", gen_str);

            data_len = strlen(data_str);
            gen_len = strlen(gen_str);
            total_len = data_len + gen_len - 1;

            for (i = 0; i < data_len; i++) data[i] = data_str[i] - '0';
            for (i = 0; i < gen_len; i++) gen[i] = gen_str[i] - '0';
            for (i = data_len; i < total_len; i++) data[i] = 0;

            printf("\n[Step 1] Data with %d zeros appended: ", gen_len - 1);
            for (i = 0; i < total_len; i++) printf("%d", data[i]);
            printf("\n");

            for (i = 0; i <= total_len - gen_len; i++) {
                if (data[i] == 1) {
                    for (j = 0; j < gen_len; j++) {
                        data[i + j] = data[i + j] ^ gen[j];
                    }
                }
            }

            printf("[Step 2] CRC Checksum (Remainder): ");
            for (i = data_len; i < total_len; i++) printf("%d", data[i]);
            printf("\n");

            printf("[Step 3] Final message to be transmitted: %s", data_str);
            for (i = data_len; i < total_len; i++) printf("%d", data[i]);
            printf("\n");
            break;

        case 2:
            printf("\n-- CRC Verification --\n");
            printf("Enter received message (data + CRC): ");
            scanf("%s", data_str);
            printf("Enter generator polynomial: ");
            scanf("%s", gen_str);

            data_len = strlen(data_str);
            gen_len = strlen(gen_str);

            for (i = 0; i < data_len; i++) data[i] = data_str[i] - '0';
            for (i = 0; i < gen_len; i++) gen[i] = gen_str[i] - '0';

            for (i = 0; i <= data_len - gen_len; i++) {
                if (data[i] == 1) {
                    for (j = 0; j < gen_len; j++) {
                        data[i + j] = data[i + j] ^ gen[j];
                    }
                }
            }

            int is_error = 0;
            printf("\n[Computation] Remainder after division: ");
            for (i = data_len - (gen_len - 1); i < data_len; i++) {
                printf("%d", data[i]);
                if (data[i] != 0) {
                    is_error = 1;
                }
            }
            printf("\n");

            if (is_error) {
                printf("\n[Result] Error detected. The remainder is not zero.\n");
            } else {
                printf("\n[Result] No error detected. The remainder is zero.\n");
            }
            break;

        default:
            printf("Invalid choice.\n");
            break;
    }
    return 0;
}

File: sem5/CN/crc.py
def crc_division(dividend, divisor):
    div_len = len(divisor)
    dividend_list = list(dividend) # Work with a mutable list

    # Loop through the dividend
    for i in range(len(dividend_list) - div_len + 1):
        # If the leading bit of the current window is 1, perform XOR
        if dividend_list[i] == '1':
            for j in range(div_len):
                # XOR operation: '1' if different, '0' if same
                dividend_list[i + j] = '0' if dividend_list[i + j] == divisor[j] else '1'
    
    # The remainder is the last (div_len - 1) bits
    remainder = "".join(dividend_list[-(div_len - 1):])
    return remainder

# --- Hardcoded Values ---
data = "10011101"
generator = "1011"
gen_len = len(generator)

print("--- SENDER ---")
print(f"Data: {data}")
print(f"Generator: {generator}")

# 1. Append (gen_len - 1) zeros to the data
appended_data = data + '0' * (gen_len - 1)
print(f"Appended Data: {appended_data}")

# 2. Calculate the CRC remainder
crc_remainder = crc_division(appended_data, generator)
print(f"Calculated CRC (Remainder): {crc_remainder}")

# 3. Create the final codeword to be sent
codeword = data + crc_remainder
print(f"Codeword to be sent: {codeword}")

# --- RECEIVER SIDE ---
print("\n--- RECEIVER ---")
print(f"Received Codeword: {codeword}")
print(f"Generator: {generator}")

# 1. Verify the received codeword by getting its remainder
remainder_at_receiver = crc_division(codeword, generator)
print(f"Remainder at receiver: {remainder_at_receiver}")

# 2. Check if the remainder is all zeros
if int(remainder_at_receiver) == 0:
    print("Result: No error detected. The data is correct.")
else:
    print("Result: Error detected! The data is corrupted.")